exchanges/deribit/deribit_websocket_eps.go

```diff
@@ -0,0 +1,2513 @@
+		if trades[x].Direction != sideBUY && trades[x].Direction != sideSELL {
+			return false, errInvalidOrderSideOrDirection
+		}
+		if trades[x].Amount <= 0 {
+			return false, errInvalidAmount
+		}
+		if trades[x].Price < 0 {
+			return false, fmt.Errorf("%w, trade price can't be negative", errInvalidPrice)
+		}
+	}
+	input := &struct {
+		Role   string            `json:"role"`
+		Trades []BlockTradeParam `json:"trades"`
+	}{
+		Role:   role,
+		Trades: trades,
+	}
+	var resp bool
+	return resp, d.SendWSRequest(matchingEPL, simulateBlockPosition, input, resp, true)
+}
+
+// SendWSRequest sends a request through the websocket connection.
+// both authenticated and public endpoints are allowed.
+func (d *Deribit) SendWSRequest(epl request.EndpointLimit, method string, params, response interface{}, authenticated bool) error {
+	if authenticated && !d.Websocket.CanUseAuthenticatedEndpoints() {
+		return errWebsocketConnectionNotAuthenticated
+	}
+	input := &WsRequest{
+		JSONRPCVersion: rpcVersion,
+		ID:             d.Websocket.Conn.GenerateMessageID(true),
+		Method:         method,
+		Params:         params,
+	}
+	resp := &wsResponse{Result: response}
+	err := d.sendWsPayload(epl, input, resp)
+	if err != nil {
+		return err
+	}
+	if resp.Error.Code != 0 || resp.Error.Message != "" {
+		var data string
+		if resp.Error.Data != nil {
+			value, err := json.Marshal(resp.Error.Data)
+			if err == nil {
+				data = string(value)
+			}
+		}
+		return fmt.Errorf("code: %d message: %s %s", resp.Error.Code, resp.Error.Message, data)
+	}
+	return nil
+}
+
+// sendWsPayload handles sending Websocket requests
+func (d *Deribit) sendWsPayload(ep request.EndpointLimit, input *WsRequest, response *wsResponse) error {
+	if input == nil {
+		return fmt.Errorf("%w, input can not be ", common.ErrNilPointer)
+	}
+	deadline := time.Now().Add(websocketRequestTimeout)
+	ctx, cancelFunc := context.WithDeadline(context.Background(), deadline)
+	defer func() {
+		if time.Now().After(deadline) {
+			cancelFunc()
+		}
+	}()
+	for attempt := 1; ; attempt++ {
+		// Initiate a rate limit reservation and sleep on requested endpoint
+		err := d.Requester.InitiateRateLimit(ctx, ep)
+		if err != nil {
+			return fmt.Errorf("failed to rate limit Websocket request: %w", err)
+		}
+
+		if d.Verbose {
+			log.Debugf(log.RequestSys, "%s attempt %d", d.Name, attempt)
+		}
+		var payload []byte
+		payload, err = d.Websocket.Conn.SendMessageReturnResponse(input.ID, input)
+		if err != nil {
+			return err
+		}
+		err = json.Unmarshal(payload, response)
+		if err != nil {
+			return err
+		}
+		switch response.Error.Code {
+		case 10040:
+			after := 100 * time.Millisecond // because all the request rate will be reset after 1 sec interval
+			backoff := request.DefaultBackoff()(attempt)
+			delay := backoff
+			if after > backoff {
+				delay = after
+			}
+			if dl, ok := ctx.Deadline(); ok && dl.Before(time.Now().Add(delay)) {
+				if err != nil {
+					return err
+				}
```

Comment:
As suggested in last review, this can be removed as the err has already been checked.

Reply:
😨  Removed!